File: src/main/java/de/hanno/hpengine/engine/model/QuadVertexBuffer.java
package de.hanno.hpengine.engine.model;

import javax.vecmath.Vector2f;
import java.util.EnumSet;

public class QuadVertexBuffer extends VertexBuffer {

    private static final VertexBuffer fullscreenBuffer = new QuadVertexBuffer(true);
    private static final VertexBuffer debugBuffer = new QuadVertexBuffer(false);
	static {
		fullscreenBuffer.upload();
		debugBuffer.upload();
	}
    public static VertexBuffer getFullscreenBuffer() {
        return fullscreenBuffer;
    }
    public static VertexBuffer getDebugBuffer() {
        return debugBuffer;
    }

	public QuadVertexBuffer(boolean fullscreen) {
		super(getPositionsAndTexCoords(fullscreen), EnumSet.of(DataChannels.POSITION3, DataChannels.TEXCOORD));
	}
	public QuadVertexBuffer(Vector2f leftBottom, Vector2f rightUpper) {
		super(getPositionsAndTexCoords(leftBottom, rightUpper), EnumSet.of(DataChannels.POSITION3, DataChannels.TEXCOORD));
	}

	static float[] getPositionsAndTexCoords(boolean fullscreen) {
		if (fullscreen) {
			return new float[] {
					-1.0f, -1.0f, 0.0f,   0f, 0f,
					1.0f, -1.0f, 0.0f,    1f, 0f,
					-1.0f,  1.0f, 0.0f,   0f,  1.0f,
					-1.0f,  1.0f, 0.0f,   0f,  1.0f,
					1.0f, -1.0f, 0.0f,    1.0f, 0f,
					1.0f,  1.0f, 0.0f,    1.0f,  1.0f
			};
		} else {
//			return new float[] {
//					-1.0f, -1.0f, 0.0f, 0f, 0f,
//					0f, -1.0f, 0.0f,    1f, 0f,
//					-1.0f,  0f, 0.0f,   0f, 1.0f,
//					-1.0f,  0f, 0.0f,   0f, 1.0f,
//					0f, -1.0f, 0.0f,    1.0f, 0f,
//					0f,  0f, 0.0f,      1.0f, 1.0f
//			};
			return getPositionsAndTexCoords(new Vector2f(-1, -1), new Vector2f(0, 0));
		}
	}
	static float[] getPositionsAndTexCoords(Vector2f leftBottom, Vector2f rightUpper) {
		return new float[] {
				leftBottom.x, leftBottom.y, 0.0f,   0f, 0f,
				rightUpper.x, leftBottom.y, 0.0f,    1f, 0f,
				leftBottom.x,  rightUpper.y, 0.0f,   0f,  1.0f,
				leftBottom.x,  rightUpper.y, 0.0f,   0f,  1.0f,
				rightUpper.x, leftBottom.y, 0.0f,    1.0f, 0f,
				rightUpper.x,  rightUpper.y, 0.0f,    1.0f,  1.0f
		};
	}
}

File: opengl/src/main/kotlin/de/hanno/hpengine/graphics/buffer/vertex/QuadVertexBuffer.kt
package de.hanno.hpengine.graphics.buffer.vertex

import de.hanno.hpengine.graphics.DataChannels
import de.hanno.hpengine.graphics.GraphicsApi
import org.joml.Vector2f
import java.util.*

private val DEFAULTCHANNELS = EnumSet.of(
    DataChannels.POSITION3,
    DataChannels.TEXCOORD,
    DataChannels.NORMAL
)

object QuadVertexBuffer {

    context(GraphicsApi)
    operator fun invoke(
        values: FloatArray = fullScreenVertices,
    ) = VertexBufferImpl(
        DEFAULTCHANNELS,
        values
    )

    val fullScreenVertices = floatArrayOf(
        -1.0f, -1.0f, 0.0f, 0f, 0f,
        1.0f, -1.0f, 0.0f, 1f, 0f,
        -1.0f, 1.0f, 0.0f, 0f, 1.0f,
        -1.0f, 1.0f, 0.0f, 0f, 1.0f,
        1.0f, -1.0f, 0.0f, 1.0f, 0f,
        1.0f, 1.0f, 0.0f, 1.0f, 1.0f
    )
    val quarterScreenVertices = getPositionsAndTexCoords(
        Vector2f(-1f, -1f),
        Vector2f(0f, 0f)
    )

    fun getPositionsAndTexCoords(leftBottom: Vector2f, rightUpper: Vector2f) = floatArrayOf(
        leftBottom.x, leftBottom.y, 0.0f, 0f, 0f,
        rightUpper.x, leftBottom.y, 0.0f, 1f, 0f,
        leftBottom.x, rightUpper.y, 0.0f, 0f, 1.0f,
        leftBottom.x, rightUpper.y, 0.0f, 0f, 1.0f,
        rightUpper.x, leftBottom.y, 0.0f, 1.0f, 0f,
        rightUpper.x, rightUpper.y, 0.0f, 1.0f, 1.0f
    )
}
